### PR TITLE
azure: refactored image gallery creation for private repos

### PR DESCRIPTION
**Pull request by pawelpros**
Azure rejects any attempts to create a private repository without set sharing profile properties which are not needed in case of private repository deployment.

### CODE REVIEW

File: azure/uploader.go [hunk 1]
```diff
@@ -337,45 +337,58 @@ func (u *Uploader) ensureManagedImageDeleted(ctx context.Context) error {
...
-	var communityGalleryInfo *armcomputev5.CommunityGalleryInfo
-	if u.config.Azure.SharingProfile == "community" {
-		communityGalleryInfo = &armcomputev5.CommunityGalleryInfo{
-			PublicNamePrefix: &pubNamePrefix,
-			Eula:             toPtr("none"),
-			PublisherContact: toPtr("test@foo.bar"),
-			PublisherURI:     toPtr("https://foo.bar"),
+
+		var communityGalleryInfo *armcomputev5.CommunityGalleryInfo
+		if u.config.Azure.SharingProfile == "community" {
```
[the remark this answers is not in the document]
Comment (pawelpros) [reply azure/uploader.go:372]: I left it as it was merged in previous PR. As I saw beside private, community there is third groups profile so in that third option galery info will not be initialized (but I haven't tested that flow)

**Comment by pawelpros** [comment]
I have used uplosi based on commit bf18a40bdfb1656d4ce95b9553074f189c50dc70 (newest one)

Config:

```
[base]
imageVersion = "0.0.1"
name = "peer-pod-0.0.1"

[base.azure]
subscriptionID = "<ID>"
location = "eastus"
resourceGroup = "<RESOURCE_GROUP>"
sharedImageGallery = "<PRIVATE_GALLERY_NAME>"
sharingProfile = "private"
imageDefinitionName = "cc-image"

[variant.default]
provider = "azure"

[variant.default.azure]
replicationRegions = ["eastus"]
```

Output:
```
/root/uplosi/uplosi upload build/system.raw
2024/12/03 08:24:26 Uploading variant default
2024/12/03 08:24:29 Deleting image version 0.0.1 in <RESOURCE_GROUP>/<PRIVATE_GALLERY_NAME>/cc-image
2024/12/03 08:26:33 Deleting managed image peer-pod-0.0.1-0.0.1 in <RESOURCE_GROUP>
2024/12/03 08:26:47 Disk peer-pod-0.0.1-0.0.1 in <RESOURCE_GROUP> doesn't exist. Nothing to clean up.
2024/12/03 08:26:48 Image gallery <PRIVATE_GALLERY_NAME> in <RESOURCE_GROUP> exists
Error: uploading variants: uploading image: ensuring sig exists: image gallery has no sharing profile
```

My PR skip additional checks against sharing profile which in my scenario is not configured in Azure at all

**Comment by katexochen** [comment]
Thanks for the details. :) So you have an already existing image gallery, that was the info I was missing.

**Comment by katexochen** [comment]
I'm still having issues to reproduce this. Did you create that existing image gallery with uplosi or in another way?
I also don't get the idea of you fix: shouldn't we change the checks we do in case there is an existing gallery instead of changing the way galleries are created?

**Comment by pawelpros** [comment]
As far as i remember (I have done it quite long time ago) i used azure CLI:

```
az sig create \
  --gallery-name "${GALLERY_NAME}" \
  --resource-group "${AZURE_RESOURCE_GROUP}" \
  --location "${AZURE_REGION}"

export GALLERY_IMAGE_DEF_NAME="cc-image"
az sig image-definition create \
  --resource-group "${AZURE_RESOURCE_GROUP}" \
  --gallery-name "${GALLERY_NAME}" \
  --gallery-image-definition "${GALLERY_IMAGE_DEF_NAME}" \
  --publisher GreatPublisher \
  --offer GreatOffer \
  --sku GreatSku \
  --os-type "Linux" \
  --os-state "Generalized" \
  --hyper-v-generation "V2" \
  --location "${AZURE_REGION}" \
  --architecture "x64" \
  --features SecurityType=ConfidentialVmSupported
```

Hmm another approach maybe better would be conditionally check these if `u.config.Azure.SharingProfile != "private"`:

```
if resp.Gallery.Properties.SharingProfile == nil {
  return errors.New("image gallery has no sharing profile")
}
if resp.Gallery.Properties.SharingProfile.Permissions == nil {
  return errors.New("image gallery has no sharing profile permissions")
}
if *resp.Gallery.Properties.SharingProfile.Permissions != *sharingProf {
  return errors.New("image gallery has different sharing profile permissions, cannot update automatically")
}
```

**Comment by katexochen** [comment]
@pawelpros I've created https://github.com/edgelesssys/uplosi/pull/96 to fix this issue, could you take a look?

**Pull request closed without merging**